src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceDao.java

```diff
@@ -0,0 +1,424 @@
+            .addValue("profile_id", application.getProfileId());
+        DaoKeyHolder keyHolder = new DaoKeyHolder();
+        jdbcTemplate.update(sql, params, keyHolder);
+        return keyHolder.getId();
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
+    public AzureApplicationDeploymentResource retrieveApplicationDeploymentById(UUID id) {
+        MapSqlParameterSource params = new MapSqlParameterSource().addValue("id", id);
+        return retrieveApplicationDeploymentBy(sqlApplicationRetrieveById, params);
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
+    public AzureApplicationDeploymentResource retrieveApplicationDeploymentByName(String applicationName) {
+        MapSqlParameterSource params =
+            new MapSqlParameterSource().addValue("azure_application_deployment_name", applicationName);
+        return retrieveApplicationDeploymentBy(sqlApplicationRetrieveByDepName, params);
+    }
+    @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
+    public AzureApplicationDeploymentResource retrieveApplicationDeploymentByIdForDelete(UUID id) {
+        MapSqlParameterSource params = new MapSqlParameterSource().addValue("id", id);
+        return retrieveApplicationDeploymentBy(sqlApplicationRetrieveByIdForDelete, params);
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
+    public List<AzureApplicationDeploymentResource> retrieveApplicationDeploymentsByBillingProfileId(
+        UUID billingProfileId) {
+        MapSqlParameterSource params = new MapSqlParameterSource().addValue("profile_id", billingProfileId);
+        return retrieveApplicationDeploymentListBy(sqlApplicationRetrieveByBillingProfileId, params);
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public List<UUID> markUnusedApplicationDeploymentsForDelete(UUID profileId) {
+        // Note: this logic is copied over from the project logic in GCP and may be overkill for the azure usecase but
+        // this will make it less likely to accidentally delete storage accounts
+        // Collect all application deployments related to the incoming profile and compute the number of references
+        // on those apps.
+        MapSqlParameterSource params = new MapSqlParameterSource().addValue("profile_id", profileId);
+        List<AppRefs> appRefs = jdbcTemplate.query(sqlProfileProjectRefs, params,
+            (rs, rowNum) -> new AppRefs()
+                .projectId(rs.getObject("aid", UUID.class))
+                .refCount(rs.getLong("refcnt")));
+
+        // If the profile is in use by any project, we bail here.
+        long totalRefs = 0;
+        for (AppRefs ref : appRefs) {
+            logger.info("Profile app reference appId: {} refCount: {}", ref.getAppId(), ref.getRefCount());
+            totalRefs += ref.getRefCount();
+        }
```

Comment:
Take it or leave it, but Streams are pretty cool!

```suggestion
        long totalRefs = appRefs.stream()
            .peek(r -> logger.info("Profile app reference appId: {} refCount: {}", r.getAppId(), r.getRefCount()))
            .mapToLong(AppRefs::getRefCount).sum();
        }
```

Reply:
Yeah, I agree but I want to be careful not to go to overboard with lambdas...it can hurt readability a little bit

Reply:
I actually find lambdas to be more readable in a lot of cases. Especially with streams, the operations are more sequential and don't modify a variable that I have to keep track of in my head. But these are stylistic choices and not necessary for merging

Reply:
Yeah, I don't disagree with you. Maybe we can start leaning on them more as we move forward